geom/validation_test.go

```diff
@@ -195,18 +196,23 @@ func TestMultiPolygonValidation(t *testing.T) {
 func BenchmarkPolygonSingleRingValidation(b *testing.B) {
 	for _, sz := range []int{10, 100, 1000, 10000} {
 		b.Run(fmt.Sprintf("n=%d", sz), func(b *testing.B) {
-			coords := [][]Coordinates{{}}
-			coords[0] = make([]Coordinates, sz+1)
+			floats := make([]float64, 2*(sz+1))
 			for i := 0; i < sz; i++ {
 				angle := float64(i) / float64(sz) * 2 * math.Pi
-				coords[0][i].X = math.Cos(angle)
-				coords[0][i].Y = math.Sin(angle)
+				floats[2*i+0] = math.Cos(angle)
+				floats[2*i+1] = math.Sin(angle)
+			}
+			floats[2*sz+0] = floats[0]
```

Comment:
Nit: The new float sequence technique feels like a reduction in (type) safety (as compared with using more structures types). That doesn't invalidate it, but I'm just noticing the code is harder to read and feel confident about.

Reply:
Agreed, there are two aspects where we lose type safety:

- These is less safely in the size of the floats slice. There is nothing to enforce that the length is a multiple of the coordinate type dimension. The code (purposefully and in a documented manner) panics in this case. Even though it's documented, that's definitely a con.
- It's less clear explicit which elements of the slice are Xs and Ys (and Ms and Zs).

In some sense, it does _improve_ type safety (for some aspects of the code). The main alternative that I considered was to store a slice of `Coordinates` rather than a slice of `float64`s. If we were to continue to keep track of the coordinates type within the `Coordinates` type, then these could potentially differ within a sequence (and we would have to check for that and give an error). This is actually similar to the way that rings work in polygons in this change (which I'm not too happy about...). The other more obvious disadvantage to that approach is that we're storing Z and M values even if they're not used.